File: react-native/__tests__/reportService.test.ts
import { ReportService } from '../src/services/reportService';
import { Logger } from '../src/core/logger';
import { OfflineQueue } from '../src/core/offlineQueue';
import { LogLevel } from '../src/models/enums';
import type { ReportCreate, ReportDetail, ReportList } from '../src/models/report';

// Mock AsyncStorage
jest.mock('@react-native-async-storage/async-storage', () => ({
  getItem: jest.fn(() => Promise.resolve(null)),
  setItem: jest.fn(() => Promise.resolve()),
  removeItem: jest.fn(() => Promise.resolve()),
}));

// Mock NetInfo
jest.mock('@react-native-community/netinfo', () => ({
  addEventListener: jest.fn(() => jest.fn()),
  fetch: jest.fn(() =>
    Promise.resolve({ isConnected: true, isInternetReachable: true }),
  ),
}));

const sampleDetail: ReportDetail = {
  id: 'rpt-001',
  plateNumber: 'ABC1234',
  vehicleTypeId: 'vt-car',
  vehicleTypeName: 'Car',
  infractionId: 'inf-001',
  infractionName: 'Illegal Parking',
  severity: 'medium' as any,
  description: 'Double parked on main street',
  status: 'submitted' as any,
  location: { latitude: 40.4168, longitude: -3.7038, address: 'Test St' },
  occurredAt: '2025-01-15T10:00:00Z',
  createdAt: '2025-01-15T10:05:00Z',
  updatedAt: '2025-01-15T10:05:00Z',
  reporter: { id: 'user-001', fullName: 'Test User', email: 'test@example.com' } as any,
  evidence: [],
  verificationCount: 0,
  rejectionCount: 0,
};

const sampleList: ReportList = {
  items: [
    {
      id: 'rpt-001',
      plateNumber: 'ABC1234',
      infractionName: 'Illegal Parking',
      severity: 'medium' as any,
      status: 'submitted' as any,
      location: { latitude: 40.4168, longitude: -3.7038 },
      occurredAt: '2025-01-15T10:00:00Z',
      createdAt: '2025-01-15T10:05:00Z',
      evidenceCount: 0,
    },
  ],
  total: 1,
  page: 1,
  pageSize: 20,
  totalPages: 1,
};

describe('ReportService', () => {
  let service: ReportService;
  let mockHttp: any;
  let logger: Logger;
  let offlineQueue: OfflineQueue;

  beforeEach(() => {
    mockHttp = {
      get: jest.fn(),
      post: jest.fn(),
      put: jest.fn(),
      delete: jest.fn(),
    };
    logger = new Logger(LogLevel.None);
    offlineQueue = new OfflineQueue(logger, false);
    service = new ReportService(mockHttp, offlineQueue, logger);
  });

  describe('create', () => {
    it('sends POST /reports and returns ReportDetail', async () => {
      mockHttp.post.mockResolvedValue({ data: sampleDetail });

      const report: ReportCreate = {
        plateNumber: 'ABC1234',
        vehicleTypeId: 'vt-car',
        infractionId: 'inf-001',
        description: 'Double parked',
        location: { latitude: 40.4168, longitude: -3.7038 },
      };

      const result = await service.create(report);

      expect(mockHttp.post).toHaveBeenCalledWith('/reports', report);
      expect(result).toEqual(sampleDetail);
    });

    it('returns the result from the API', async () => {
      mockHttp.post.mockResolvedValue({ data: sampleDetail });

      const result = await service.create({
        plateNumber: 'XYZ',
        vehicleTypeId: 'vt-car',
        infractionId: 'inf-002',
        description: 'Test',
        location: { latitude: 0, longitude: 0 },
      });

      expect((result as ReportDetail).id).toBe('rpt-001');
    });
  });

  describe('list', () => {
    it('sends GET /reports and returns ReportList', async () => {
      mockHttp.get.mockResolvedValue({ data: sampleList });

      const result = await service.list();

      expect(mockHttp.get).toHaveBeenCalledWith('/reports', { params: undefined });
      expect(result.items).toHaveLength(1);
      expect(result.total).toBe(1);
    });

    it('passes pagination parameters', async () => {
      mockHttp.get.mockResolvedValue({ data: sampleList });

      await service.list({ page: 2, pageSize: 10 });

      expect(mockHttp.get).toHaveBeenCalledWith('/reports', {
        params: { page: 2, pageSize: 10 },
      });
    });

    it('passes status filter', async () => {
      mockHttp.get.mockResolvedValue({ data: sampleList });

      await service.list({ status: 'verified' });

      expect(mockHttp.get).toHaveBeenCalledWith('/reports', {
        params: { status: 'verified' },
      });
    });
  });

  describe('getById', () => {
    it('sends GET /reports/:id and returns ReportDetail', async () => {
      mockHttp.get.mockResolvedValue({ data: sampleDetail });

      const result = await service.getById('rpt-001');

      expect(mockHttp.get).toHaveBeenCalledWith('/reports/rpt-001');
      expect(result.id).toBe('rpt-001');
      expect(result.plateNumber).toBe('ABC1234');
    });
  });

  describe('getByPlate', () => {
    it('sends GET /reports/by-plate/:plate', async () => {
      mockHttp.get.mockResolvedValue({ data: sampleList });

      await service.getByPlate('ABC1234');

      expect(mockHttp.get).toHaveBeenCalledWith('/reports/by-plate/ABC1234');
    });

    it('encodes special characters in plate', async () => {
      mockHttp.get.mockResolvedValue({ data: sampleList });

      await service.getByPlate('AB CD');

      expect(mockHttp.get).toHaveBeenCalledWith('/reports/by-plate/AB%20CD');
    });
  });

  describe('delete', () => {
    it('sends DELETE /reports/:id', async () => {
      mockHttp.delete.mockResolvedValue({});

      await service.delete('rpt-001');

      expect(mockHttp.delete).toHaveBeenCalledWith('/reports/rpt-001');
    });
  });
});

File: flutter/multando_sdk/test/report_service_test.dart
import 'dart:convert';

import 'package:dio/dio.dart';
import 'package:flutter_test/flutter_test.dart';
import 'package:multando_sdk/multando_sdk.dart';
import 'package:multando_sdk/src/core/http_client.dart';
import 'package:multando_sdk/src/core/auth_manager.dart';

/// A [HttpClientAdapter] that returns canned responses for testing.
class _MockAdapter implements HttpClientAdapter {
  final Map<String, ResponseBody Function(RequestOptions)> _handlers = {};

  void onGet(String path, ResponseBody Function(RequestOptions) handler) {
    _handlers['GET:$path'] = handler;
  }

  void onPost(String path, ResponseBody Function(RequestOptions) handler) {
    _handlers['POST:$path'] = handler;
  }

  void onDelete(String path, ResponseBody Function(RequestOptions) handler) {
    _handlers['DELETE:$path'] = handler;
  }

  @override
  Future<ResponseBody> fetch(
    RequestOptions options,
    Stream<List<int>>? requestStream,
    Future<void>? cancelFuture,
  ) async {
    final key = '${options.method}:${options.path}';
    // Try exact match first, then prefix match for parameterized paths.
    final handler = _handlers[key] ??
        _handlers.entries
            .where((e) => key.startsWith(e.key))
            .map((e) => e.value)
            .firstOrNull;
    if (handler != null) {
      return handler(options);
    }
    return ResponseBody.fromString(
      jsonEncode({'detail': 'Not Found'}),
      404,
      headers: {
        Headers.contentTypeHeader: ['application/json'],
      },
    );
  }

  @override
  void close({bool force = false}) {}
}

ResponseBody _jsonBody(dynamic data, {int status = 200}) {
  return ResponseBody.fromString(
    jsonEncode(data),
    status,
    headers: {
      Headers.contentTypeHeader: ['application/json'],
    },
  );
}

void main() {
  group('ReportService', () {
    late Dio dio;
    late _MockAdapter adapter;
    late ReportService service;

    final sampleReportDetail = {
      'id': 'rpt-001',
      'infraction_id': 'inf-123',
      'plate_number': 'ABC1234',
      'location': {
        'latitude': 40.4168,
        'longitude': -3.7038,
        'address': 'Test Street',
      },
      'status': 'submitted',
      'source': 'sdk',
      'reporter_id': 'user-001',
      'created_at': '2025-01-15T10:00:00Z',
      'evidence': [],
    };

    final sampleReportList = {
      'items': [
        {
          'id': 'rpt-001',
          'infraction_id': 'inf-123',
          'plate_number': 'ABC1234',
          'status': 'submitted',
          'source': 'sdk',
          'created_at': '2025-01-15T10:00:00Z',
        },
      ],
      'total': 1,
      'page': 1,
      'page_size': 20,
      'total_pages': 1,
    };

    setUp(() {
      adapter = _MockAdapter();
      dio = Dio(BaseOptions(
        baseUrl: 'https://api.test.multando.io/api/v1',
      ));
      dio.httpClientAdapter = adapter;

      // Create a minimal MultandoHttpClient with the dio instance injected.
      // Since we need the _http field, we create a ReportService directly.
      final httpClient = MultandoHttpClient(
        config: const MultandoConfig(
          baseUrl: 'https://api.test.multando.io',
          apiKey: 'test-key',
        ),
        authManager: AuthManager(),
      );
      // Override the internal dio with our mock adapter.
      httpClient.dio.httpClientAdapter = adapter;

      service = ReportService(httpClient: httpClient);
    });

    test('create() sends POST /reports and returns ReportDetail', () async {
      adapter.onPost('/reports', (_) => _jsonBody(sampleReportDetail));

      final report = ReportCreate(
        infractionId: 'inf-123',
        plateNumber: 'ABC1234',
        location: const LocationData(latitude: 40.4168, longitude: -3.7038),
      );

      final result = await service.create(report);
      expect(result, isNotNull);
      expect(result!.id, equals('rpt-001'));
      expect(result.plateNumber, equals('ABC1234'));
      expect(result.status, equals(ReportStatus.submitted));
    });

    test('list() sends GET /reports and returns ReportList', () async {
      adapter.onGet('/reports', (_) => _jsonBody(sampleReportList));

      final result = await service.list();
      expect(result.items, hasLength(1));
      expect(result.total, equals(1));
      expect(result.page, equals(1));
      expect(result.items.first.plateNumber, equals('ABC1234'));
    });

    test('list() sends pagination parameters', () async {
      adapter.onGet('/reports', (options) {
        expect(options.queryParameters['page'], equals(2));
        expect(options.queryParameters['page_size'], equals(10));
        return _jsonBody(sampleReportList);
      });

      await service.list(page: 2, pageSize: 10);
    });

    test('getById() sends GET /reports/:id and returns ReportDetail', () async {
      adapter.onGet('/reports/rpt-001', (_) => _jsonBody(sampleReportDetail));

      final result = await service.getById('rpt-001');
      expect(result.id, equals('rpt-001'));
      expect(result.infractionId, equals('inf-123'));
    });

    test('delete() sends DELETE /reports/:id', () async {
      var deleteCalled = false;
      adapter.onDelete('/reports/rpt-001', (_) {
        deleteCalled = true;
        return _jsonBody({'ok': true});
      });

      await service.delete('rpt-001');
      expect(deleteCalled, isTrue);
    });
  });
}
